### PR TITLE
Fix loss tensor data contract

### PR DESCRIPTION
**Pull request by jdblackstar**
## Summary
- Make loss `forward` methods accept Tensor or ndarray inputs, unwrap to NumPy, and return scalar NumPy data for Tensor wrapping.
- Keep `backward(context, grad)` aligned with autograd by returning Tensor gradients and `None` for targets.
- Update loss tests to exercise the operation-style contract, including raw-logit cross entropy with one-hot targets through `Tensor.cross_entropy`.

## Tests
- `.venv/bin/python -m pytest tests/test_loss.py -q`

---

> [!NOTE]
> **Medium Risk**
> Training math changed for cross-entropy (batch mean and 1D handling) and all losses now depend on the shared input contract, so any caller passing wrong shapes or expecting old scalar-sum CE could break silently.
> 
> **Overview**
> Loss `forward`/`backward` now go through shared helpers so inputs can be **`Tensor` or `ndarray`**, upstream grads are normalized consistently, and **`forward` returns scalar NumPy** for the autograd wrapper.
> 
> **Cross-entropy family** changes are the behavioral ones: **`_classification_arrays`** promotes 1D logits/targets to a batch row, validates shapes, and restores the original layout in **`backward`**. **`CrossEntropyLoss`** and **`CategoricalCrossEntropyLoss`** now **mean over the batch** (and scale gradients by batch size), which updates expected values versus the old 1D path. **MSE** and **binary CE** follow the same unwrap/`_grad_data` pattern without changing their reduction semantics beyond test fixes.
> 
> **`tests/test_loss.py`** now drives losses through **`Function` + `backward(context, grad)`**, checks Tensor and raw array inputs, and adds an end-to-end **`Tensor.cross_entropy` → `backward()`** test.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit f247f1e019e5047a22d7852e949d9a2069df0819. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by jdblackstar** [comment]
closing, superseded by #14
